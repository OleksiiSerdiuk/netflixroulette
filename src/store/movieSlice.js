import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import API from '../api';

export const initialState = {
  movies: null,
  isLoading: false,
  serverError: null,
  filters: ['all', 'documentary', 'comedy', 'horror', 'crime'],
  sortBy: 'release_date',
  filter: 'all',
  searchField: '',
};

export const fetchMovies = createAsyncThunk(
  '/movies/fetchMovies',
  async (property, { rejectWithValue, getState }) => {
    const {
      movies: { sortBy, filter, searchField },
    } = getState();

    const currentFilter = filter === 'all' ? '' : filter;

    try {
      const response = await API.get(
        `/movies?sortBy=${sortBy}&sortOrder=asc&filter=${currentFilter}&limit=9&searchBy=title&search=${searchField}`
      );

      return await response.data;
    } catch (err) {
      return rejectWithValue(err.response);
    }
  }
);

export const removeMovie = createAsyncThunk(
  '/movies/removeMovies',
  async (movieId, { rejectWithValue }) => {
    try {
      await API.delete(`/movies/${movieId}`);
    } catch (err) {
      return rejectWithValue(err.response);
    }
  }
);

export const createMovie = createAsyncThunk(
  '/movies/createMovie',
  async (data, { rejectWithValue }) => {
    try {
      await API.post(`/movies`, data);
    } catch (err) {
      return rejectWithValue(err.response);
    }
  }
);

export const updateMovie = createAsyncThunk(
  '/movies/updateMovie',
  async (data, { rejectWithValue }) => {
    try {
      await API.put(`/movies`, data);
    } catch (err) {
      return rejectWithValue(err.response);
    }
  }
);

export const movieSlice = createSlice({
  name: 'movies',
  initialState,
  reducers: {
    setFilter: (state, action) => {
      state.filter = action.payload;
    },
    setSort: (state, action) => {
      state.sortBy = action.payload;
    },
    setSearchField: (state, action) => {
      state.searchField = action.payload;
    },
  },
  extraReducers: {
    [fetchMovies.pending]: (state) => {
      state.isLoading = true;
    },
    [fetchMovies.fulfilled]: (state, action) => {
      state.movies = action.payload.data;
      state.isLoading = false;
    },
    [fetchMovies.rejected]: (state, action) => {
      state.serverError = action.payload.data;
      state.isLoading = false;
    },
  },
});

export const { setFilter, setSort, setSearchField } = movieSlice.actions;

export default movieSlice.reducer;
