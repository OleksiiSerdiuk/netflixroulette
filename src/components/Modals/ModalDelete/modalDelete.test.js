import { fireEvent, render, waitFor } from '@testing-library/react';
import { ModalDelete } from './index';
import store from '../../../store';
import { Provider } from 'react-redux';
import { act } from 'react-dom/test-utils';

const handleSubmit = jest.fn();

describe('modal delete', () => {
  test('renders modal delete snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <ModalDelete />
      </Provider>
    );
    expect(asFragment(<ModalDelete />)).toMatchSnapshot();
  });

  it('should handle submit', async () => {
    const { getByRole } = render(
      <Provider store={store}>
        <ModalDelete />
      </Provider>
    );

    await act(() => {
      fireEvent.click(getByRole('button', { type: /submit/i }));
      handleSubmit();
    });
    await waitFor(() => {
      expect(handleSubmit).toHaveBeenCalledTimes(1);
    });
  });
});
