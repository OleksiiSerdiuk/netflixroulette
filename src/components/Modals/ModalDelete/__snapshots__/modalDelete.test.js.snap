// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modal delete renders modal delete snapshot 1`] = `
<DocumentFragment>
  <div
    class="section-modal-delete"
  >
    <p
      class="section-modal-delete__text"
    >
      Are you sure you want to delete this movie?
    </p>
    <button
      class="btn btn-outline btn-outline-corral btn-align-right"
      type="submit"
    >
      Confirm
    </button>
  </div>
</DocumentFragment>
`;
